### PR TITLE
Add microbenchmarks for Crc32 and Crc64 algorithms

### PR DESCRIPTION
**Pull request by brantburnett**
See https://github.com/dotnet/runtime/pull/86539#issuecomment-1557172918

### CODE REVIEW

**Comment by brantburnett** [comment]
@adamsitnik 

Here are the CRC benchmarks per your request. I do see an execution failure in one case in CI, but my attempts to make sense of it have hit a dead end. I don't think it's related to this change, but can't be certain.

**Comment by danmoseley** [comment]
Yea, it's not your change, some fragility in a winforms test.

**Review by danmoseley — APPROVED** [review]

**Review by danmoseley — APPROVED** [review]

File: src/benchmarks/micro/libraries/System.IO.Hashing/CrcAppendTests.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        public void Setup()
+        {
+            Buffer = ValuesGenerator.Array<byte>(BufferSize);
+            Crc = new T();
+        }
+
+        [Benchmark]
+        public void Append()
+        {
+            Crc.Append(Buffer);
```
Comment (adamsitnik) [inline src/benchmarks/micro/libraries/System.IO.Hashing/CrcAppendTests.cs:30]: Please excuse me if this is a dumb question (I've become an owner of `System.IO.Hashing` quite recently), but is my understanding correct that `Append` is not having any side-effects that could prolong the execution time of the next invocations and we don't need to call `Crc.Reset()` for every benchmark?

File: src/benchmarks/micro/libraries/System.IO.Hashing/CrcAppendTests.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+// See the LICENSE file in the project root for more information.
+
+using BenchmarkDotNet.Attributes;
+using BenchmarkDotNet.Extensions;
+using MicroBenchmarks;
+
+namespace System.IO.Hashing.Tests
+{
+    [BenchmarkCategory(Categories.Libraries)]
+    public abstract class CrcAppendTests<T>
```
Comment (adamsitnik) [inline src/benchmarks/micro/libraries/System.IO.Hashing/CrcAppendTests.cs:12]: in the future we might add benchmarks for other methods, so we could make the type name more general
```suggestion
    public abstract class Crc_Perf<T>
```

File: src/benchmarks/micro/libraries/System.IO.Hashing/Crc64AppendTests.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using BenchmarkDotNet.Attributes;
+
+namespace System.IO.Hashing.Tests
+{
+    public class Crc64AppendTests : CrcAppendTests<Crc64>
```
Comment (adamsitnik) [inline src/benchmarks/micro/libraries/System.IO.Hashing/Crc64AppendTests.cs:9]:
```suggestion
    public class Crc64_Perf : Crc_Perf<Crc64>
```

File: src/benchmarks/micro/libraries/System.IO.Hashing/Crc32AppendTests.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using BenchmarkDotNet.Attributes;
+
+namespace System.IO.Hashing.Tests
+{
+    public class Crc32AppendTests : CrcAppendTests<Crc32>
```
Comment (adamsitnik) [inline src/benchmarks/micro/libraries/System.IO.Hashing/Crc32AppendTests.cs:9]:
```suggestion
    public class Crc32_Perf : Crc_Perf<Crc32>
```

File: src/benchmarks/micro/libraries/System.IO.Hashing/CrcAppendTests.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+            Buffer = ValuesGenerator.Array<byte>(BufferSize);
+            Crc = new T();
+        }
+
+        [Benchmark]
+        public void Append()
+        {
+            Crc.Append(Buffer);
+        }
+    }
```
Comment (adamsitnik) [inline src/benchmarks/micro/libraries/System.IO.Hashing/CrcAppendTests.cs:32]: Would you mind adding benchmarks for other methods? Like [[GetCurrentHash()](https://learn.microsoft.com/en-us/dotnet/api/system.io.hashing.noncryptographichashalgorithm.getcurrenthash?view=dotnet-plat-ext-7.0#system-io-hashing-noncryptographichashalgorithm-getcurrenthash)](https://learn.microsoft.com/en-us/dotnet/api/system.io.hashing.noncryptographichashalgorithm.getcurrenthash?view=dotnet-plat-ext-7.0#system-io-hashing-noncryptographichashalgorithm-getcurrenthash(system-span((system-byte)))) for example?

**Review by adamsitnik — COMMENTED** [review]
@brantburnett thank you for your contribution! Overall it looks good, but PTAL at my comments.

**Review by adamsitnik — COMMENTED** [review]
@brantburnett thank you for your contribution! Overall it looks good, but PTAL at my comments.

**Comment by cincuranet** [comment]
@adamsitnik is there anything that prevents this from merging?

**Comment by cincuranet** [comment]
I'm going to merge this as it is now and add the remaining benchmarks myself in separate PR.

**Pull request merged**